substratevm/src/com.oracle.svm.core/src/com/oracle/svm/core/heap/GCName.java

```diff
@@ -0,0 +1,89 @@
+/*
+ * Copyright (c) 2022, 2022, Oracle and/or its affiliates. All rights reserved.
+ * Copyright (c) 2022, 2022, Red Hat Inc. All rights reserved.
+ * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
+ *
+ * This code is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 only, as
+ * published by the Free Software Foundation.  Oracle designates this
+ * particular file as subject to the "Classpath" exception as provided
+ * by Oracle in the LICENSE file that accompanied this code.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+package com.oracle.svm.core.heap;
+
+import java.util.ArrayList;
+import org.graalvm.nativeimage.Platform;
+import org.graalvm.nativeimage.Platforms;
+import org.graalvm.nativeimage.hosted.Feature;
+
+import com.oracle.svm.core.annotate.AutomaticFeature;
+import com.oracle.svm.core.annotate.Uninterruptible;
+
+/**
+ * This class holds supported garbage collector names.
+ */
+public class GCName {
+    @Platforms(Platform.HOSTED_ONLY.class) private static final ArrayList<GCName> HostedGCNameList = new ArrayList<>();
+
+    protected static GCName[] GCNames;
+
+    private final int id;
+    private final String name;
+
+    @Platforms(Platform.HOSTED_ONLY.class)
+    protected GCName(String name) {
+        this.name = name;
+        this.id = addGCNameMapping();
+    }
+
+    @Platforms(Platform.HOSTED_ONLY.class)
+    private int addGCNameMapping() {
+        synchronized (HostedGCNameList) {
+            int newId = HostedGCNameList.size();
+            HostedGCNameList.add(newId, this);
+            GCNames = HostedGCNameList.toArray(new GCName[HostedGCNameList.size()]);
```

Comment:
I don't think that this line is necessary as this is already handled by the feature below.
If you run into issues after removing the line (e.g., that the value is always null at run-time for some reason), then it might be necessary to annotate the field `GCNames` with `@UnknownObjectField(types= {GCName[].class})`

Reply:
Ah, I was puzzled why it works for `GCCause `and I kept getting `null `in `jfrTypeRepository.writeGCNames()`

Reply:
But after change, I got:
```   
object: com.oracle.svm.core.genscavenge.GenScavengeGCName@7d8d671b  of class: com.oracle.svm.core.genscavenge.GenScavengeGCName
  reachable through:
    object: [Lcom.oracle.svm.core.heap.GCName;@6235ff82  of class: com.oracle.svm.core.heap.GCName[]
    root: HostedField<GCName.GCNames location: 648   AnalysisField<GCName.GCNames accessed: false reads: true written: true folded: false>>

	at com.oracle.svm.core.util.VMError.shouldNotReachHere(VMError.java:72)
	at com.oracle.svm.hosted.NativeImageGenerator.doRun(NativeImageGenerator.java:675)
	at com.oracle.svm.hosted.NativeImageGenerator.run(NativeImageGenerator.java:512)
	at com.oracle.svm.hosted.NativeImageGeneratorRunner.buildImage(NativeImageGeneratorRunner.java:423)
	at com.oracle.svm.hosted.NativeImageGeneratorRunner.build(NativeImageGeneratorRunner.java:598)
	at com.oracle.svm.hosted.NativeImageGeneratorRunner.main(NativeImageGeneratorRunner.java:128)
	at com.oracle.svm.hosted.NativeImageGeneratorRunner$JDK9Plus.main(NativeImageGeneratorRunner.java:628)
```

Reply:
I am going to fix this issue when I am merging this PR to master.

Reply:
I had to do a few refactorings on top, to make everything work. This PR will be merged to master in the next few days.